game-core/src/main/java/games/strategy/triplea/attachments/UnitAttachment.java

```diff
@@ -1817,20 +1817,22 @@ private void resetBlockade() {
 
   private void setGivesMovement(final String value) throws GameParseException {
     final String[] s = splitOnColon(value);
-    if (s.length <= 0 || s.length > 2) {
+    if (s.length < 2) {
       throw new GameParseException(
-          "givesMovement cannot be empty or have more than two fields" + thisErrorMsg());
-    }
-    final String unitTypeToProduce = s[1];
-    // validate that this unit exists in the xml
-    final UnitType ut = getData().getUnitTypeList().getUnitType(unitTypeToProduce);
-    if (ut == null) {
-      throw new GameParseException("No unit called:" + unitTypeToProduce + thisErrorMsg());
+          "givesMovement must have an integer followed by 1 or more unit types" + thisErrorMsg());
+    }
+    final int movement = getInt(s[0]);
+    for (int i = 1; i < s.length; i++) {
+      final String unitTypeName = s[i];
+      // validate that this unit exists in the xml
+      final UnitType type = getData().getUnitTypeList().getUnitType(unitTypeName);
+      if (type == null) {
```

Comment:
Yeah, if you pass an invalid unit type then it returns null. So if in the XML I specify say "infantr" instead of "infantry" then it would return null. I'm not really a big fan of that syntax as you are literally creating an optional to then immediately check it which to me isn't really any better than checking the null.

Now you could argue to update the getUnitType method to return an optional which I think does provide value but is also probably a fairly large refactoring.

Reply:
> syntax as you are literally creating an optional to then immediately check it which to me isn't really any better than checking the null.

That's by design, it's a functional way to do the same rather than imperative.

> Now you could argue to update the getUnitType method to return an optional which I think does provide value but is also probably a fairly large refactoring.

Yes, but the refactoring gets easier if the callers use optionals

Reply:
Agree. Personally I think I prefer the imperative over the functional syntax in this case but probably depends on how much you use functional java syntax.

Reply:
Sure, java has started to lean towards functional style. If anything, it removes the keyword 'null'. Part of it is existing and ingrained habits and idioms compared to new. For a while, iterator loops were easier to understand than foreach loops, simply because of habits. Once something is really ingrained, the understanding becomes automatic and takes no effort to understand, seeing something new, but perhaps better, takes some work.

But again, removing the null keyword is of value, we'd make refactoring easier, and the code is more direct in assigning the value or failing to do so, more atomic. The reduces logic flow generally makes it easier to understand as there are few conditionals and the statements are more direct to their purpose.